gnovm/pkg/gnolang/coverage.go

```diff
@@ -0,0 +1,642 @@
+	io commands.IO,
+) error {
+	scanner := bufio.NewScanner(file)
+	lineNumber := 1
+
+	for scanner.Scan() {
+		line := scanner.Text()
+		hitCount, covered := coverage.HitLines[lineNumber]
+
+		lineInfo := c.formatLineInfo(
+			lineNumber,
+			line,
+			hitCount,
+			covered,
+			coverage.ExecutableLines[lineNumber],
+			showHits,
+		)
+		io.Printfln(lineInfo)
+
+		lineNumber++
+	}
+
+	return scanner.Err()
+}
+
+func (c *CoverageData) formatLineInfo(
+	lineNumber int,
+	line string,
+	hitCount int,
+	covered, executable, showHits bool,
+) string {
+	lineNumStr := fmt.Sprintf("%4d", lineNumber)
+	color := c.color(covered, executable)
+	hitInfo := c.getHitInfo(hitCount, covered, showHits)
+
+	format := "%s%s%s %s%s%s%s"
+
+	return fmt.Sprintf(format, color, lineNumStr, colorReset, hitInfo, color, line, colorReset)
+}
+
+func (c *CoverageData) color(covered, executable bool) string {
+	switch {
+	case covered:
+		return colorGreen
+	case executable:
+		return colorYellow
+	default:
+		return colorWhite
+	}
+}
+
+func (c *CoverageData) getHitInfo(hitCount int, covered, showHits bool) string {
+	if !showHits {
+		return ""
+	}
+
+	if covered {
+		return fmt.Sprintf("%s%-4d%s ", colorOrange, hitCount, colorReset)
+	}
+
+	return strings.Repeat(" ", 5)
+}
+
+func (c *CoverageData) findMatchingFiles(pattern string) []string {
+	var files []string
+	for file := range c.Files {
+		if strings.Contains(file, pattern) {
+			files = append(files, file)
+		}
+	}
+	return files
+}
+
+// floor1 round down to one decimal place
+func floor1(v float64) float64 {
+	return math.Floor(v*10) / 10
+}
+
+func getCoverageColor(percentage float64) string {
+	switch {
+	case percentage >= 80:
+		return colorGreen
+	case percentage >= 50:
+		return colorYellow
+	default:
+		return colorRed
+	}
+}
+
+func calculateCoverage(covered, total int) float64 {
+	return float64(covered) / float64(total) * 100
+}
+
+// findAbsoluteFilePath finds the absolute path of a file given its relative path.
+// It starts searching from root directory and recursively traverses directories.
+func (c *CoverageData) findAbsoluteFilePath(filePath string) (string, error) {
+	c.mu.RLock()
+	cachedPath, ok := c.pathCache[filePath]
+	c.mu.RUnlock()
+	if ok {
+		return cachedPath, nil
+	}
+
+	var result string
+	err := filepath.WalkDir(c.rootDir, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if !d.IsDir() && strings.HasSuffix(path, filePath) {
+			result = path
+			return filepath.SkipAll
+		}
+		return nil
+	})
+	if err != nil {
+		return "", err
+	}
+
+	if result == "" {
+		return "", fmt.Errorf("file %s not found", filePath)
+	}
+
+	c.mu.Lock()
+	c.pathCache[filePath] = result
+	c.mu.Unlock()
+
+	return result, nil
+}
+
+func isTestFile(pkgPath string) bool {
+	return strings.HasSuffix(pkgPath, "_test.gno") ||
+		strings.HasSuffix(pkgPath, "_testing.gno") ||
+		strings.HasSuffix(pkgPath, "_filetest.gno")
+}
+
+type jsonCoverageMap map[string]JSONFileCoverage
```

Comment:
You should move these types to the `ToJSON` method, and not have them be package-level types